server/src/main/java/org/elasticsearch/action/search/RankFeaturePhase.java

```diff
@@ -59,16 +77,130 @@ public void onFailure(Exception e) {
         });
     }
 
-    private void innerRun() throws Exception {
-        // other than running reduce, this is currently close to a no-op
+    void innerRun() throws Exception {
+        // if the RankBuilder specifies a QueryPhaseCoordinatorContext, it will be called as part of the reduce call
+        // to operate on the first `window_size * num_shards` results and merge them appropriately.
         SearchPhaseController.ReducedQueryPhase reducedQueryPhase = queryPhaseResults.reduce();
-        moveToNextPhase(queryPhaseResults, reducedQueryPhase);
+        RankFeaturePhaseRankCoordinatorContext rankFeaturePhaseRankCoordinatorContext = coordinatorContext(context.getRequest().source());
+        if (rankFeaturePhaseRankCoordinatorContext != null) {
+            ScoreDoc[] queryScoreDocs = reducedQueryPhase.sortedTopDocs().scoreDocs(); // rank_window_size
+            final List<Integer>[] docIdsToLoad = SearchPhaseController.fillDocIdsToLoad(context.getNumShards(), queryScoreDocs);
+            final CountedCollector<SearchPhaseResult> rankRequestCounter = new CountedCollector<>(
+                rankPhaseResults,
+                context.getNumShards(),
+                () -> onPhaseDone(rankFeaturePhaseRankCoordinatorContext, reducedQueryPhase),
+                context
+            );
+
+            // we send out a request to each shard in order to fetch the needed feature info
+            for (int i = 0; i < docIdsToLoad.length; i++) {
+                List<Integer> entry = docIdsToLoad[i];
+                SearchPhaseResult queryResult = queryPhaseResults.getAtomicArray().get(i);
+                if (entry == null || entry.isEmpty()) {
+                    if (queryResult != null) {
+                        releaseIrrelevantSearchContext(queryResult, context);
```

Comment:
Maybe I misunderstand how this works, but doesn't the `fetch` phase run after this `rank` phase? 

Is it possible for us to release a search context here, but then that context becomes relevant for the fetch phase and thus we released it too early?

I am only worried as this code seems copy-pasted from `fetch` but `fetch` is the final phase, and this `rank` phrase is NOT the final phase.

Reply:
I need to think about it a bit more, but the idea here was that since we trim-down results from the query phase (since query phase returned `n` * `rank_window_size` results from `n` shards, but we reduce to just `rank_window_size` from `m` shards, where `m<=n`), it's possible that  when reaching `FetchSearchPhase` and perform the same operation there through:
```
                    SearchPhaseResult queryResult = queryResults.get(i);
                    if (entry == null) { // no results for this shard ID
                        if (queryResult != null) {
```
we could have some contexes that were never closed (as `queryResult` will be null)
A couple of things to note: 
* `queryResults` is a really bad & misleading name with the current set of changes, as it could be `rankFeatureResults` (passed from `RankFeaturePhase#onPhaseDone` ). Will update this & add some comments.
* the `shard` result could be `null` if after computing the top `rank_window_size` results (through `SearchPhaseController#reducedQueryPhase`) we skipped some shards and never reached out to retrieve additional field data. 

So, in `rank` requests, the part in `FetchSearchPhase` is expected to close the search contexes for the shards where: 
* some docs were present in the top `rank_window_size` results in the `RankFeaturePhase` (and this the context was not cleared there) and we do have a `RankFeatureResult`
* are not part of the top `size` docs for which we want to execute `fetch`,  so we need to close the leftover resources. 

It's a bit confusing and tricky, and I'll need to revisit this to make sure, but does the above make sense? WDYT?

Reply:
Your explaining it makes sense. 

Do we have a multi-node test where the `size` for fetch is smaller than the window_size for the reranking? We should make sure that we exercise this path.

Reply:
We do have the `FieldBasedRerankerIT#testFieldBasedRerankerPagination` which uses a `rank_window_size` of `10` and a `size` of `2`, so depending on the setup (as there is some randomness there) we could very well end up meeting all the scenarios above.  Do you think that this is enough, or should I try to add something explicitly for to verify the above?

Reply:
@pmpailis explicit testing is better than random for these strange edge cases that we know about .

Reply:
Added a test (`testNotAllShardsArePresentInFetchPhase`) in b3cdfe670d60229d06bf3716c92f0da1b59c9899 that: 
* indexes some documents using different routing keys
* specifies `rank_window_size` > `size`
* at least one shard does not contribute  to `RankFeaturePhase` and we call the `releaseIrrelevantSearchContext`
* at least one shard does not contribute to `FetchSearchPhase` and we call the `releaseIrrelevantSearchContext`

In addition to that, I've also updated `releaseIrrelevantSearchContext` to a non-static, protected method.